main/pipeline/pipeline.cc

```diff
@@ -78,25 +78,66 @@ class CFGCollectorAndTyper {
     }
 };
 
-string fileKey(core::GlobalState &gs, core::FileRef file) {
-    auto path = file.data(gs).path();
+string fileKey(const core::File &file, string_view subtype = ""sv) {
+    auto path = file.path();
     string key(path.begin(), path.end());
     key += "//";
-    auto hashBytes = sorbet::crypto_hashing::hash64(file.data(gs).source());
+    if (!subtype.empty()) {
+        key += subtype;
+        key += "//";
+    }
+    auto hashBytes = sorbet::crypto_hashing::hash64(file.source());
     key += absl::BytesToHexString(string_view{(char *)hashBytes.data(), size(hashBytes)});
     return key;
 }
 
-unique_ptr<ast::Expression> fetchTreeFromCache(core::GlobalState &gs, core::FileRef file,
-                                               const unique_ptr<KeyValueStore> &kvstore) {
-    if (kvstore && file.id() < gs.filesUsed()) {
-        string fileHashKey = fileKey(gs, file);
+string fileHashKey(const core::File &file) {
+    return fileKey(file, "filehash");
+}
+
+unique_ptr<const core::FileHash> fetchFileHashFromCache(spd::logger &logger, core::File &file,
+                                                        const unique_ptr<OwnedKeyValueStore> &kvstore) {
+    if (kvstore) {
+        string key = fileHashKey(file);
+        auto maybeCached = kvstore->read(key);
+        if (maybeCached) {
+            prodCounterInc("types.input.filehashes.kvstore.hit");
+            file.cachedFileHash = true;
+            return core::serialize::Serializer::loadFileHash(logger, maybeCached);
+        } else {
+            prodCounterInc("types.input.filehashes.kvstore.miss");
+        }
+    }
+    return nullptr;
+}
+
+void cacheFileHashes(const unique_ptr<OwnedKeyValueStore> &kvstore, const vector<shared_ptr<core::File>> &files) {
```

Comment:
Aren't these already stored in Files of GlobalState(togather with the AST), why store them separately?

Reply:
> Aren't these already stored in Files of GlobalState(togather with the AST)

That's what I thought, but they aren't. :( The AST isn't stored with a file. It's stored under a key created using a hash of the file's contents.

When Sorbet serializes the GlobalState into the cache, it only serializes payload files.

We don't serialize non-payload files into the cache. If I were to do that, I'm sure the cache would balloon significantly for large projects.